### PR TITLE
test: pin every register-resolver matching clause with a tracked fixture tree (#346)

### LINKED ISSUE #346: Duplication-register resolver: two matching clauses survive mutation with no subcheck
Found by the cross-family review lens on PR #345 (#329); out of that PR's scope, filed per standing practice.

## Problem

`tests/test_docs_registry.py` resolves register citations with an ordered list of definition patterns. Two clauses in that resolver can be weakened with no test going red:

- The `python_constant` branch requires the constant to start at column zero (`lines[line_number].startswith(segment)`). Removing that clause leaves the suite green, so an indented assignment of the same name inside a function would resolve as the module constant.
- `HEADING_PATTERN` (`^#{1,6}\s+{H}\s*$`) can be widened to `^.*$` with the suite still green, because no subcheck asserts that a heading citation fails against a non-heading line containing the same text.

No current register citation resolves wrongly; the risk is a future citation resolving to the wrong definition or line.

## Fix shape

Add two rows to the test's existing subcheck tables rather than new tests: a Python constant name that exists only indented in a tracked file (expected `unresolved symbol ...`), and a heading citation whose text appears in the Markdown file only as body prose (expected `heading not found`). Verify each by the mutation above going red.

### PR DESCRIPTION
**Pull request by leehopper**
Closes #346

## What changed

The register-citation resolver's matching clauses had no test that flips when one is weakened; the cross-family review of PR #345 (#329) found two, and a census over every clause found 48 of 53 mutants surviving.

The diff changes ten paths: the resolver test, `pyproject.toml`, `.pre-commit-config.yaml`, and seven fixture files.

- Added `tests/fixtures/citation_resolver/`, seven tracked data files with one positive line per accepted form and one near-miss line per narrowing constraint (`.py`, `.js`, `.mjs`, `.yaml`, `.yml`, node:test titles, Markdown headings).
- Added `TestDocsRegistry::test_duplication_resolver_matching_clauses`: 129 span and expected-reason rows, plus 17 rows derived from the suffix map (one foreign-kind line per definition fixture and kind its suffix does not grant, which must exist and must not resolve). Each definition fixture declares the kinds its suffix grants on its first line, and the test compares that declaration with the map.
- Added witness-based assertions: every `DEFINITION_PATTERNS` index and every witnessed kind is reached by a positive row, one witness per resolved segment, every map suffix has a fixture file, and every fixture file has a positive row.
- Added keyword-only witness recording to `_definition_reason`, `_title_reason`, and `_citation_reason`.
- Deleted the dead Python constant `startswith` check and changed heading matching to `heading_regex.search`.
- Hoisted the suffix map to `DEFINITION_KINDS_BY_SUFFIX`, the heading suffix to `HEADING_SUFFIX` and the title kind to `TITLE_KIND`, so the test derives its coverage sets from the resolver's own constants.

## Why the shape

The fixture tree is tracked because its bytes are resolver input. Each row's trailing comment names the matching clause it pins.

The test loops over the rows, then asserts the coverage relationships. Every definition pattern index and every witnessed kind needs a positive witness, and headings need one witness.

The suffix map's keys must equal the definition fixtures' suffixes, and its values are pinned by the derived rows: widening any suffix's kind set makes a foreign line resolve.

The optional keyword-only `witness=None` channel records `(kind, pattern_index, line_number)` after successful matches.

Existing callers remain unchanged. Results matched for every span in `docs/duplication-register.md` with and without a witness list.

The heading pattern's `^` is now its only start anchor, matching every other pattern's convention. No other behavior changes.

Scope ruling: pin every clause in one PR rather than the two the issue named; no paired bench run.

## Mutation ledger

The initial census covered 53 matching-clause mutants and found 48 survivors.

The final ledger contains 191 mutants. It includes the census, issue probes, and sweeps of every pattern's optional groups, whitespace classes, and alternations.

It also covers the `_title_reason` suffix gate, the fourteen single-pair widenings of the suffix map's values, the class delimiter arms, the heading and title suffix gates, the node:test paired delimiter, one compound tab narrowing per pattern, first-pattern-wins, four witness-fidelity mutants, and three structural probes.

The structural probes append a dummy `DEFINITION_PATTERNS` entry, add `.ts` to the suffix map, and drop a kind from the foreign-symbol table. Each fails a derived assertion rather than a row.

Each mutant fails the new test. The builder independently verified this with a driver and witness harness.

Opus re-verified the plan, and a luna mutation lens re-ran the ledger in a fresh worktree with restoration proven after each round. Round two added the map-value rows after the Opus branch lens found them missing; round three pinned the class delimiter arms and the fixture kind declarations; round four pinned the heading branch's Markdown-suffix gate; round five pinned the title gate's mjs arm, paired delimiters, tab gaps and the heading witness index; round six pinned first-pattern-wins and the YAML comment gap.

A prior row already kills widening `HEADING_PATTERN` to `^.*$`.

The issue-named rows are:

- `A3_drop_caret_only` widens the constant anchor to `^[ \t]*`, changing `definitions.py::INDENTED` from `unresolved symbol 'INDENTED' after line 0` to `None`.
- `B2_heading_contains` changes `HEADING_PATTERN` to `^.*{H}.*$`, changing `headings.md#Embedded Heading` from `heading not found` to `None`.

The deleted `startswith` check was equivalent because `^` already anchors column zero under `re.search` on a `splitlines()` element.

Declared equivalences are:

- Reordering `DEFINITION_PATTERNS` does not change a return value, and witness coverage is set-based.
- `.search` versus `.match` remains equivalent when the heading or title pattern keeps `^`.
- Dropping `^` while switching to `.match` is also equivalent in the heading and title branches.
- `re.escape(delimiter)` is equivalent because the three delimiters contain no regular-expression syntax.
- `[^=\n]` and `[^=]` are equivalent after `splitlines()` removes newline characters from annotation input.
- The derived rows' `symbol not in text` guard and the first-line kinds declaration are deliberately redundant: a widened map value shows up in two places of a diff, and neither guard can be pinned by a test-file mutant on its own.
- Widening the heading suffix gate from equality to a prefix test survives because no `.mdx` or `.mdown` file is tracked.
- Narrowing one `[ \t]` class of a pattern to `[ ]` while a neighbouring class absorbs the tab is equivalent; the per-pattern compound (every class narrowed) is what the tab lines pin.

Unpinned by convention: `if found is not None` versus `if found` in the cursor loop differs only for a definition on a file's first line that recurs later with a child between; every definition fixture puts its purpose comment and kinds declaration on that line, so no row covers it.

## Lint and CI

Fixture data is excluded from Ruff and mypy using their configured path exclusions.

Markdownlint-fix and trailing-whitespace also exclude the tree because its bytes are test input; end-of-file-fixer keeps its guard.

That includes two trailing spaces on one heading line. YAML fixtures remain valid for check-yaml, and Biome lints `workflows/` only.

`pre-commit run --all-files` passes 15/15 on every commit. CI is green on every commit (`afa8565` 16/16; `cc1b83f` through `8053865` 17/17).

## Adjacent

The resolver does not track fenced code blocks, so `CONTRIBUTING.md#Pipeline pytest suite` resolves to `None` although its only match is a shell comment inside a fence. Out of scope here; filed as #348.

## Review

The plan was drafted by Codex Astra, red-teamed by one Claude Opus and one Codex Astra pass, revised, and re-verified by Opus.

Codex luna built it. Three luna lenses (mutation ledger, plan conformance, docs and lint) reviewed every round; an Opus branch lens reviewed rounds one through five.

Lee's own code-gauntlet run follows.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0128A82yNUwSENjdJCgcA9Dj

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/bugbot) is generating a summary for commit 8053865a5abbc8ac99152b60a57f6ae4714bd552. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/test_docs_registry.py [hunk 1]
```diff
@@ -830,6 +844,488 @@ def add_error(line, span, reason):
...
+                expected_count = {"file": 0, "symbol": len(symbols)}.get(
+                    citation_kind, 1
+                )
+                if len(witness) != expected_count:
+                    add_error(
+                        span,
+                        f"{len(witness)} witnesses recorded, expected {expected_count}",
+                    )
+                witnessed.extend(witness)
+                positive_paths.add(_parse_citation(span)[1])
```
Comment (leehopper) [inline tests/test_docs_registry.py]: **💡 [LOW] _parse_citation(span) is called twice per positive row when one call suffices**

Inside the citations loop, _parse_citation(span) is called once and its path is thrown away with an underscore, then called again three lines later purely to recover that same path for positive_paths.add(). _parse_citation is a pure function of span, so capturing path in the first destructure removes the second call entirely. Before: citation_kind, _, symbols, _ = _parse_citation(span) ... positive_paths.add(_parse_citation(span)[1]). After: citation_kind, path, symbols, _ = _parse_citation(span) ... positive_paths.add(path).

**Suggested fix:**
Capture path on the first call and reuse it: citation_kind, path, symbols, _ = _parse_citation(span) then positive_paths.add(path), deleting the second call.

```suggestion
                citation_kind, path, symbols, _ = _parse_citation(span)
                expected_count = {"file": 0, "symbol": len(symbols)}.get(
                    citation_kind, 1
                )
                if len(witness) != expected_count:
                    add_error(
                        span,
                        f"{len(witness)} witnesses recorded, expected {expected_count}",
                    )
                witnessed.extend(witness)
                positive_paths.add(path)
```

⚔️ *Code Gauntlet*

**Review by leehopper — COMMENTED** [review]
### ⚔️ Code Gauntlet

This PR (714 changed lines) claims to add golden fixture files exercising `tests/test_docs_registry.py`'s citation resolver — a self-test used to validate that documentation citations in the repo (definition symbols, node-test titles, and Markdown headings) actually resolve against real source — and to refactor the resolver's matching helpers so each match records a "witness" (kind, pattern index) that the new test can assert against. New fixtures are added under `tests/fixtures/citation_resolver/`: `definitions.py`, `definitions.js`, `definitions.mjs`, `definitions.yaml`, `definitions.yml` (positive and near-miss forms for each `DEFINITION_PATTERNS` regex, including tab-whitespace and cross-language "foreign" symbol variants), `titles.js` (node `test`/`describe`/`it` title-matching edge cases), and `headings.md` (Markdown heading matching edge cases). `tests/test_docs_registry.py` is modified to introduce `TITLE_KIND`, `HEADING_SUFFIX`, and a hoisted `DEFINITION_KINDS_BY_SUFFIX` map, thread an optional `witness` list through `_definition_reason`, `_title_reason`, and `_citation_reason` to record which pattern/kind matched, change heading matching from `re.match`+`any()` to an explicit per-line loop using `re.search`, remove a redundant `lines[line_number].startswith(segment)` check for `python_constant`, and add a large new test (`test_duplication_resolver_matching_clauses`) enumerating ~130 citation-span/expected-reason pairs plus consistency assertions (foreign-symbol coverage per kind, fixture-declared vs. map-granted kinds, witness index/kind completeness, and positive-row coverage of every fixture file). `.pre-commit-config.yaml` and `pyproject.toml` are updated to exclude the new `tests/fixtures/citation_resolver/` fixtures from `trailing-whitespace`, `markdownlint-fix`, ruff, and mypy, since the fixtures intentionally contain non-conforming bytes (tabs, trailing whitespace, malformed syntax) as test data. The stated motivation is to pin/regression-test the citation resolver's matching behavior (issue/PR #346 per commit history) so that future changes to its regex patterns or matching logic are caught structurally rather than by prose review; risk is concentrated in the resolver refactor and its new assertions in `tests/test_docs_registry.py` (test logic, not a fixture, classified medium risk), while the fixture files themselves are marked low risk and skip-listed as frozen golden data by the repo's own REVIEW.md, and the tooling-config changes are narrowly scoped exclusion-pattern additions.

1 finding(s) after the gauntlet — 1 low. 1 routed as improvement suggestion(s).

---
Generated by code-gauntlet | Reviewed up to: 8053865a5abbc8ac99152b60a57f6ae4714bd552

File: tests/test_docs_registry.py (hunk 1 above)
Comment (leehopper) [reply tests/test_docs_registry.py]: Fixed: the first destructure now captures `path` and the second call is gone.

File: tests/fixtures/citation_resolver/definitions.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+def scope():
+    INDENTED = 1
+python_key: 1
+def early(): pass
+class Parent:
+    def child(self): pass
+class Parent: pass
+async def py_async(): pass
+classJoined:
+asyncdef async_joined(): pass
```
[the remark this answers is not in the document]
Comment (leehopper) [reply tests/fixtures/citation_resolver/definitions.py:25]: Intentional. `asyncdef` is a near-miss fixture line: the resolver must not match it as an async definition. The fixture tree is test data, excluded from ruff and mypy, and never imported.

File: tests/fixtures/citation_resolver/definitions.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+def scope():
+    INDENTED = 1
+foreign_js: 1
+def early(): pass
+class Parent:
+    def child(self): pass
+class Parent: pass
+async def py_async(): pass
+classJoined:
+asyncdef async_joined(): pass
```
[the remark this answers is not in the document]
Comment (leehopper) [reply tests/fixtures/citation_resolver/definitions.py:25]: Intentional. `asyncdef` is a near-miss fixture line: the resolver must not match it as an async definition. The fixture tree is test data, excluded from ruff and mypy, and never imported.

File: tests/fixtures/citation_resolver/definitions.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
```
[the remark this answers is not in the document]
Comment (leehopper) [reply tests/fixtures/citation_resolver/definitions.py:25]: Intentional. `asyncdef` is a near-miss fixture line: the resolver must not match it as an async definition. The fixture tree is test data, excluded from ruff and mypy, and never imported.

**Comment by leehopper** [comment]
@liatrio-software-factory Approve
